### PR TITLE
[#24] Add feature to make layers accessable by name and not index

### PR DESCRIPTION
**Pull request by T00mm**
I've tried my hand for #23, but this will break projects using array-indexes. If a less drastic feature is requested we could maybe just make a function named `getLayer(name)` (maybe I should have started with that)

### CODE REVIEW

**Comment by kamranayub** [comment]
As discussed IRL with the team, I think we could make this a non-BC by supporting both string and number-based indexes.

File: src/ITiledMap.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@
 export interface ITiledMap {
    width: number;
    height: number;
-   layers: ITiledMapLayer[];
+   layers: { [id: string] : ITiledMapLayer };
```
Comment (kamranayub) [inline src/ITiledMap.ts]: I think we could support the previous method of accessing layers and the new way. See suggestions.

```suggestion
   layers: { [id: string | number] : ITiledMapLayer };
```

File: src/index.ts [hunk 2]
```diff
@@ -125,7 +125,12 @@ export default class TiledResource extends Resource<ITiledMap> {
       if (data === void 0) {
          throw `Tiled map resource ${this.path} is empty`;
       }
-
+      // Change data.layers from array to a dictionary for easier access
+      var tmpLayers = <any>data.layers
+      data.layers = {};
+      for(var layer of tmpLayers){
```
Comment (kamranayub) [inline src/index.ts]:
```suggestion
      for (let i = 0; i < tmpLayers.length; i++) {
        data.layers[i] = layer;
        data.layers[layer.name] = layer;
      }
```

File: src/index.ts [hunk 3]
```diff
@@ -156,19 +161,20 @@ export default class TiledResource extends Resource<ITiledMap> {
...
-         if (layer.type === "tilelayer") {
-            for (var i = 0; i < layer.data.length; i++) {
-               let gid = <number>layer.data[i];
-
-               if (gid !== 0) {
-                  var ts = this.getTilesetForTile(gid);
-
-                  map.data[i].sprites.push(new TileSprite(ts.firstgid.toString(), gid - ts.firstgid))
+      for (var layerId in this.data.layers) {
+         if(this.data.layers.hasOwnProperty(layerId)){
```
Comment (kamranayub) [inline src/index.ts]:
```suggestion
         if(this.data.layers.hasOwnProperty(layerId) && typeof layerId === "number"){
```

File: src/index.ts [hunk 4]
```diff
@@ -182,22 +188,24 @@ var parseJsonMap = (data: ITiledMap): ITiledMap => {
 
    // Decompress layers
    if (data.layers) {
-      for (var layer of data.layers) {
-
-         if (typeof layer.data === "string") {
+      for (var layerId in data.layers) {
+         if(this.data.layers.hasOwnProperty(layerId)){
```
Comment (kamranayub) [inline src/index.ts]:
```suggestion
         if(this.data.layers.hasOwnProperty(layerId) && typeof layerId === "number"){
```

**Comment by kamranayub** [comment]
Actually my suggestions would also result in a BC since anyone iterating over `layers` would be broken by this.

If we want to introduce a BC, a better one might be to do something like this:

```ts
export interface ITiledMap {
  layers: { all: ITiledMapLayer[], byName: { [key: string]: ITiledMapLayer } }
}
```

Then layers can be accessed efficiently either way, by index or by name. It would also be a less drastic BC, causing people to only need to use `layers.all` to their calls instead of `layers`.

**Comment by eonarheim** [comment]
@T00mm I like @kamranayub's suggestion on a way to avoid a breaking change thusly. We'll need to tack another property onto the layers array and build the byName lookup

```ts
export interface ITiledMap {
  layers: ITiledMapLayer[] & { byName: {[key: string]: ITiledMapLayer }
}
```

This would allow:

```ts
map.layers[2].name // "walls"
map.layers.byName.walls
map.layers.byName[map.layers[3].name]
```

**Comment by eonarheim** [comment]
Hi @T00mm I took a stab at what we are thinking

https://www.typescriptlang.org/play/index.html#src=%0D%0A%0D%0Ainterface%20ITiledMapLayer%20%7B%0D%0A%20%20%20%20name%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20ITiledMap%20%7B%0D%0A%20%20%20%20layers%3A%20ITiledMapLayer%5B%5D%20%26%20%7B%20byName%3A%20%7B%5Bkey%3A%20string%5D%3A%20ITiledMapLayer%7D%7D%20%0D%0A%7D%0D%0A%0D%0Afunction%20BuildTiledMap()%3A%20ITiledMap%20%7B%0D%0A%20%20%20%20let%20layers%3A%20ITiledMapLayer%5B%5D%20%20%3D%20%5B%7B%20name%3A%20%22layer1%22%20%7D%2C%20%7B%20name%3A%20%22layer2%22%20%7D%2C%20%7B%20name%3A%20%22walls%22%20%7D%5D%3B%0D%0A%0D%0A%20%20%20%20%0D%0A%20%20%20%20let%20decoratedLayers%3A%20ITiledMapLayer%5B%5D%20%26%20%7B%20byName%3A%20%7B%20%5Bkey%3A%20string%5D%3A%20ITiledMapLayer%20%7D%20%7D%20%3D%20%3Cany%3Elayers%3B%0D%0A%20%20%20%20decoratedLayers.byName%20%3D%20layers.reduce((accu%2C%20curr)%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20accu%5Bcurr.name%5D%20%3D%20curr%3B%0D%0A%20%20%20%20%20%20%20%20return%20accu%3B%0D%0A%20%20%20%20%7D%2C%20%7B%7D)%3B%0D%0A%0D%0A%20%20%20%20return%20%7B%0D%0A%20%20%20%20%20%20%20%20layers%3A%20decoratedLayers%0D%0A%20%20%20%20%7D%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Alet%20map%20%3D%20BuildTiledMap()%3B%0D%0A%0D%0A%0D%0A%2F%2F%20walls%0D%0Aconsole.log(map.layers%5B2%5D.name)%0D%0Aconsole.log(map.layers.byName.walls)%0D%0Aconsole.log(map.layers.byName%5Bmap.layers%5B2%5D.name%5D)

**Pull request closed without merging**